File: LinkedListCycleII.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a list node
    def detectCycle(self, head):
        if head == None or head.next == None:
            return None
        if head.next == head:
            return head
        fast = head
        slow = head
        while True:
            if slow.next and fast.next and fast.next.next:
                slow = slow.next
                fast = fast.next.next
                if slow == fast:
                    break
            else:
                return None
        fast = head
        while fast != slow:
            slow = slow.next
            fast = fast.next
        return fast

File: LinkedListCycleII.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head || !head->next) {
            return NULL;
        }

        ListNode *fast = head, *slow = head;

        while (true) {
            if (fast->next && fast->next->next) {
                fast = fast->next->next;
                slow = slow->next;
                if (slow == fast) {
                    break;
                }
            } else {
                return NULL;
            }
        }

        fast = head;
        while(slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        return fast;
    }
};
